File: 142-linked-list-cycle-ii/linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head==None:
            return head
        s=head
        f=head
        e=head
        while f!=None and f.next!=None:
            s=s.next
            f=f.next.next
            if s==f:
                while s!=e:
                    s=s.next
                    e=e.next
                return e
        return None
        

File: 142-linked-list-cycle-ii/linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head==null || head.next==null){
            return null;
        }
        ListNode fast=head;
        ListNode slow=head;
        ListNode e=head;
        while(fast!=null && fast.next!=null){
            fast=fast.next.next;
            slow=slow.next;
            if(slow==fast){
                while(e!=slow){
                    e=e.next;
                    slow=slow.next;
                }
                return e;
            }
        }
        return null;
        
    }
}